### ISSUE TITLE
Incremental build skips building when source files deleted

### ISSUE THREAD

**Issue by bricelam** [dotnet/cli#1440]
`dotnet build` does not re-build when `*.cs` files are removed from the project.

**Comment by cdmihai** [comment]
This is a limitation of on-the-fly timestamp based incrementality.
dotnet build uses the project.json to compute the inputs and outputs on each run. It gets the input source files from the project model, which in turn gets them from the project.json file. Usually project.json projects do not explicitly reference each source file in the project (unlike csproj where you reference each source file), but rely on the implicit globbing mechanism through which cli finds source files. This means that build cannot tell when you add or delete new source files if you don't explicitly remove or put them in project.json.

In summary, with the current infrastructure, users have to either trigger a clean rebuild when they add / remove files that get implicitly discovered by CLI through globbing or explicitly put all sources in project.json (if a file gets deleted from the file system but not from project.json, dotnet build will trigger a rebuild). For CLI vNext, we should use a hash based incremental mechanism that saves IO state between consecutive build invocations.

**Comment by NTaylorMullen** [comment]
I understand the technical limitation here but this seems like a **really** poor user experience. Is there anything we can do about this for RTM?

Most of the time users sit there doing build => build => build => build rarely ever forcefully doing a rebuild.

**Comment by piotrpMSFT** [comment]
@cdmihai can you look at doing this in RTM milestone?

> For CLI vNext, we should use a hash based incremental mechanism that saves IO state between consecutive build invocations.

**Comment by piotrpMSFT** [comment]
@cdmihai is this coming in for RC2?

**Comment by cdmihai** [comment]
It should :)

**Comment by piotrpMSFT** [comment]
When do you expect it to get PR'd/merged?

**Comment by cdmihai** [comment]
After successfully rebasing on top of @pakrym's changes. Either today or tomorrow.

**Comment by piotrpMSFT** [comment]
@cdmihai it's tomorrow :) Is this ready to go?

**Comment by cdmihai** [comment]
Almost there :) 
Made a WIP PR: #2619. I'll iron out the merge issues tomorrow.

**Comment by cdmihai** [comment]
PR is ready for review / merge

**Comment by piotrpMSFT** [comment]
This was fixed by #2619
